### DISCUSSION TITLE
Initial states as a part of consensus?

### DISCUSSION THREAD

**Discussion by dahlia** [planetarium/libplanet#2750] [Ideas]
The key role of the genesis block is to found *initial states* of the chain, and we tend to say about only what states should fill it with rather than what kind of actions should be executed at first.  Also, in Libplanet, the genesis block functions as a *network identifier* as well.

In order to keep it simple and be free from generating a genesis block (which is very complicated to execute), today @longfin suggested to make initial states as a part of consensus instead of chain requiring a genesis block.  To explain it in code, we could add new properties to shape initial states:

```csharp
interface IBlockPolicy<T>
{
    // ... omitted ...
    IReadOnlyDictionary<Address, IValue> InitialStates { get; }
    IReadOnlyDictionary<Address, IReadOnlyList<FungibleAssetValue>> InitialBalances { get; }
}
```

As `BlockChain<T>` becomes to no more take any `Block<T>` via its constructors, its initial states are filled with ones from its `IBlockPolicy<T>`.

Instead, we should find an alternative to genesis hash as a network identifier.  We might add a new field to `BlockChain<T>`, or the initial state root hash may fulfill its function.

**Comment by longfin** [comment]
For reference, I've taken a look at other projects' cases.

# Ethereum
- Many implementations (like geth) seem to provide specialized ways (e.g., `genesis.json`), for sake of ease of configuration.
   - geth relies on `genesis.json` file only, and doesn't require any other datafile.
- They provide the network id and initial balances for funds. also, [it can be configurable even bytecodes and storage if needed](https://ethereum.stackexchange.com/a/30368).

# Cosmos SDK
- [Cosmos SDK also suggests the file format for genesis block](https://hub.cosmos.network/main/resources/genesis.html) like Ethereum.
- They suggest [genesis transaction concept](https://hub.cosmos.network/main/resources/genesis.html#genesis-transactions) for initial state. 

# Substrate
- They don't seem to provide a configuration file scheme for genesis, but [there is a guide](https://docs.substrate.io/reference/how-to-guides/basics/configure-genesis-state/) for setup initial state.
- According to their docs, It seems to be setting the initial state directly without going through transactions.

# NEAR Protocol
- They call the configuration for the initial state as [GenesisConfig](https://nomicon.io/GenesisConfig/).
- Interestingly, their approach seems similar to my first thought.
- So... it seems that there were also discussions about expected issues. 😅  (e.g., [1](https://github.com/near/nearcore/issues/4447), [2](https://github.com/near/nearcore/issues/3475))

**Comment by longfin** [comment]
(Just for note) 

* [leveraging system action](https://github.com/riemannulus/libplanet/pull/1) can be a sort of compromise (at least for now).
* but it means that it's too difficult to compose an API that fits harmoniously with the current structure where the genesis block is expected to maintain `.StateRootHash` for initial states... 🤔
